src/util/worker/async.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::{Arc, Mutex};
+use std::thread::{self, JoinHandle, Builder};
+use std::io;
+use std::fmt::Display;
+
+use futures::Stream;
+use futures::future::BoxFuture;
+use futures::sync::mpsc::{unbounded, UnboundedReceiver, UnboundedSender};
+use tokio_core::reactor::Core;
+
+use super::Stopped;
+
+pub trait Runnable<T: Display> {
+    // TODO: return `impl Future<Item=(), Error=()>`,
+    // once RFC: Expand and stabilize `impl Trait` is merged and implemented.
+    // See more: https://github.com/rust-lang/rfcs/pull/1951
+    fn run(&mut self, t: T) -> BoxFuture<(), ()>;
+    fn shutdown(&mut self) {}
+}
+
+/// Scheduler provides interface to schedule task to underlying workers.
+pub struct Scheduler<T> {
+    name: Arc<String>,
+    sender: UnboundedSender<Option<T>>,
+}
+
+impl<T: Display> Scheduler<T> {
+    fn new<S: Into<String>>(name: S, sender: UnboundedSender<Option<T>>) -> Scheduler<T> {
+        Scheduler {
+            name: Arc::new(name.into()),
+            sender: sender,
+        }
+    }
+
+    /// Schedule a task to run.
+    ///
+    /// If the worker is stopped, an error will return.
+    pub fn schedule(&self, task: T) -> Result<(), Stopped<T>> {
+        debug!("scheduling task {}", task);
+        if let Err(err) = self.sender.send(Some(task)) {
+            return Err(Stopped(err.into_inner().unwrap()));
+        }
+        Ok(())
+    }
+}
+
+impl<T: Display> Clone for Scheduler<T> {
```

Comment:
It seems can't, rustc blames `the method `clone` exists but the following trait bounds were not satisfied: T : std::clone::Clone`